Embeddings book search example

gawi151 opened the pull request:
Example of book search using embeddings and chat APIs. 
Example book is `Twenty Thousands Leagues Under the Sea`.

Resolves #10 

main screen:
![image](https://github.com/nomtek/flutter_ai_examples/assets/2642942/5b293425-d82f-4850-9e82-94fc456ba578)

book fragments dialog:
![image](https://github.com/nomtek/flutter_ai_examples/assets/2642942/12738508-1f51-4996-b267-d7a9a23a4c57)


todo:
- [x] make ui more responsive and looking at least decent
- [x] ~~highlight the searched phrases in the found text~~
- [x] add a prompt for searching for answers in the found text

gawi151 committed:
add example book to assets to do a search

gawi151 committed:
prepare routing for book search example

gawi151 committed:
wip

gawi151 committed:
Merge remote-tracking branch 'origin/main' into embeddings_book_search_example

gawi151 committed:
add characters and langchain deps

gawi151 committed:
add generated embeddings and tokens for example book fragments

gawi151 committed:
fix mistral tokenizer to not depend on Flutter

gawi151 committed:
Add text chunk splitter class

gawi151 committed:
add models for search data and result

gawi151 committed:
add script to generate SearchData

gawi151 committed:
add readme for Book search example

gawi151 committed:
regenerate models.dart

gawi151 committed:
initial book search ui with search logic

gawi151 committed:
add tile tile to home screen with navigation to book search example

gawi151 committed:
Refactor code

- move algebra functions to separate file
- rename BookTextSearch to BookSearch
- introduce Answer class
- show number of fragment next to it
- redesign how results are shown to user
- add more complex logic for finding answers using chat and prompts
- add todos to resolve in next commit

Co-authored-by: MarcinGruchala <MarcinGruchala@users.noreply.github.com>

gawi151 committed:
fix comment

gawi151 committed:
revert api key

gawi151 committed:
rename things in BookSearch class to be more readable

gawi151 committed:
refactor few bits in MistralAIBookSearchPage

gawi151 committed:
add comments to models in book example

gawi151 committed:
move text fragment splitter into prepare_data script

gawi151 committed:
add comment to FragmentSimilarityList class

gawi151 committed:
add description for BookSearch

gawi151 committed:
remove redundant comment

gawi151 committed:
Merge remote-tracking branch 'origin/main' into embeddings_book_search_example

gawi151 committed:
update book

gawi151 committed:
Fix bug in calculating embeddings for book fragments when sending batches to Mistral API

- update dependencies versions
- add comments with explaining things

MarcinGruchala commented on lib/mistral_ai_book_search_example/README.md:
My suggestions for what you can add to README:
- instruction on how to add a new book 
- description of the prompt and logic we use to create the answer to the user questions
 - a demo file as a gif or mp4 showing how the example works

gawi151 replied:
Good points. I will add those sections to the README. Thx 👍

gawi151 replied:
I updated the README in https://github.com/nomtek/flutter_ai_examples/pull/7/commits/4d1a2720c54abeadf13aa7cf18915fac3e7ff865

MarcinGruchala commented on lib/mistral_tokenizer/mistral_tokenizer.dart:
```diff
@@ -170,7 +174,7 @@ class MistralTokenizer {
             // if it does happen, let's follow the js implementation
             // and add <UNK> token instead of crash
-            debugPrint(
+            print(
```
Why we are using `print` instead of `debugPrint`? If we want to keep this print I suggest using lint ignore on it

gawi151 replied:
Oh that's a mistake. Thx for catching it. I'll change it back to debugPrint.

gawi151 replied:
Fixed in https://github.com/nomtek/flutter_ai_examples/pull/7/commits/2eea6ff4be95f983b20e512d48c0edd2c75f56b8

MarcinGruchala reviewed (commented):
I think that we can update the README later when we work on the README issue. But `print` in `mistral_tokenizer.dart` can create problems with our lint.

gawi151 committed:
Merge branch 'main' into embeddings_book_search_example

gawi151 committed:
Fix problem with print lint in mistral tokenizer

gawi151 committed:
rename book_text_search.dart to book_search.dart

gawi151 committed:
add demo mp4 for book search example

gawi151 committed:
add more docs to prepare_data script

gawi151 committed:
update README

- add sections about how we get answers
- add section with description and demo video
- add section about how to use different book
- add features of the example

MarcinGruchala reviewed (dismissed):
LGTM

gawi151 committed:
Fix README markdown format

- add gif to show it in readme

gawi151 committed:
make gif smaller

gawi151 committed:
smaller gif
